### PR TITLE
Normalizing covariance factors and creating functions for factor plots

### PR DESCRIPTION
**Pull request by andrewram4287**
Creating functions to plot covariance and mean factors. Incorporated normalization for covariance factors

### CODE REVIEW

File: gmm/tensor.py [hunk 1]
```diff
@@ -184,6 +184,19 @@ def get_factors_xarray(self, X):
     def norm_NK(self) -> jnp.ndarray:
         """Normalizes NK values to percentages"""
         return self.nk / np.sum(self.nk)
+    
+    def norm_covariances(self):
+        """Normalizes covarainces factors"""
+        for i in range(len(self.covFacs)):
+            self.covFacs[i] = self.covFacs[i]/np.linalg.norm(self.covFacs[i],axis=0)
```
Comment (aarmey) [inline gmm/tensor.py:191]: It looks like this removes the scaling, but would it be good to keep this around somewhere? (In the CP tensor class, these are stored as `weights`.)

File: gmm/tensor.py [hunk 2]
```diff
@@ -184,6 +184,19 @@ def get_factors_xarray(self, X):
...
+
+        covarSignal = np.empty([self.covars.shape[0], self.covars.shape[1], self.covars.shape[2]])
+        for i in range(self.covars.shape[2]):
+            sigsigcovar = self.covars[:, :, i] @ self.covars[:, :, i].T
+            covarSignal[:,:,i] = sigsigcovar 
+        covarSignal = covarSignal/np.linalg.norm(covarSignal, axis=(0,1))
```
Comment (aarmey) [inline gmm/tensor.py:197]: Same question about `weights` here.

File: gmm/tensor.py [hunk 3]
```diff
@@ -184,6 +184,19 @@ def get_factors_xarray(self, X):
     def norm_NK(self) -> jnp.ndarray:
         """Normalizes NK values to percentages"""
         return self.nk / np.sum(self.nk)
+    
+    def norm_covariances(self):
```
Comment (aarmey) [inline gmm/tensor.py:188]: When is this run? If this doesn't materially affect the results, should we perhaps run this at the end of fitting?

It would be good to come up with a test to check that this does not affect anything...

**Review by aarmey — CHANGES_REQUESTED** [review]
A few questions.

File: gmm/figures/figure6.py [hunk 4]
```diff
@@ -195,11 +86,69 @@ def make_blob_tensor(blob_DF):
...
+            sns.heatmap(data=XA[DimCol[i]], xticklabels=XA[DimCol[i]].coords[XA[DimCol[i]].dims[1]].values,
+                    yticklabels=XA[DimCol[i]].coords[XA[DimCol[i]].dims[0]].values, ax=ax[i + 1], vmin=-1,vmax=1)
+            ax[i + 1].tick_params(axis='y', rotation=0)
+            ax[i + 1].set_title('Mean Factors')
+            
+    return
+
+def plotCovFactors(fac, facXA, DimCol, n_cluster, ax):
+    """Plots covarinace factors for dimension, time, and cluster""" 
+    covarDim, covarSig = fac.norm_covariances()
```
Comment (andrewram4287) [inline gmm/figures/figure6.py:131]: @aarmey This where I am normalizing the covariances, but I think you are right that maybe having it be the last step in the analysis would make more sense

Comment (aarmey) [reply gmm/figures/figure6.py:131]: Yeah—you can keep your function as-is, but just run it at the end of the fitting.

File: gmm/tests/test_GMM.py [hunk 5]
```diff
@@ -202,4 +203,29 @@ def test_cov_fit():
...
-    np.testing.assert_allclose(cov, covR, atol=0.1, rtol=0.01)
\ No newline at end of file
+    np.testing.assert_allclose(cov, covR, atol=0.1, rtol=0.01)
+    
+def test_cov_diff():
+    """Test that ensure the orginal covariance factors are unchanged"""
+    fit, _, _ = minimize_func(data_import, rank=3, n_cluster=5, maxiter=20, seed=1)
+    fac = fit[0]
+            
+    TestCase.assertNotEqual(fac.covFacs[0],fac.covarianceFactors[0], "Error")
```
Comment (aarmey) [inline gmm/tests/test_GMM.py:213]: I'm confused. You want to test whether they are unchanged, but your assertion here is that they are not equal? Wouldn't they be equal if unchanged? Also, the built covariances should be unchanged, but the factors should be changed.

Also, this is the wrong function for arrays. Use numpy's assert_allclose: https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose

File: gmm/tests/test_GMM.py [hunk 6]
```diff
@@ -202,4 +203,29 @@ def test_cov_fit():
...
+    fit, _, _ = minimize_func(data_import, rank=4, n_cluster=6, maxiter=20, seed=1)
+    fac = fit[0]
+    for i in range(len(fac.covFacs)):
+        fac.covFacs[i] = fac.covarianceFactors[i]
+        
+    fac.covars = fac.covarianceSig
+    
+    newCovFacs, _, newCovSig, _ = fac.norm_covariances()
+    
+    assert np.testing.assert_allclose(fac.covFacs[0],newCovFacs[0],rtol=0.01)
```
Comment (aarmey) [inline gmm/tests/test_GMM.py:229]: Some of the above applies here, too.

**Review by aarmey — CHANGES_REQUESTED** [review]

**Comment by andrewram4287** [comment]
@aarmey both tests aren't working, but I believe my implementation to rebuild the tensors and redo the normalization is correct.

**Comment by aarmey** [comment]
@andrewram4287 I've fixed the tests here. I think all that needs updating are the figures. Note that I've applied the weights to just one of the factors, so there is no separate weights storage anymore.

**Comment by andrewram4287** [comment]
@aarmey  `test_cov_fit` is not passing the current test because the weights are put into the `covars` Also, when we build the covariance 3-d tensor for the signals, do we rebuild the covariance the same way as before, using `covars @ covars.T`, or just `covars`?

**Comment by aarmey** [comment]
@andrewram4287 `test_cov_fit` also fails occasionally on master, too. Let's just remove it. (Not a great test.)

Reconstructing the covariance matrices is exactly the same as before.

**Comment by andrewram4287** [comment]
> @andrewram4287 `test_cov_fit` also fails occasionally on master, too. Let's just remove it. (Not a great test.)
> 
> Reconstructing the covariance matrices is exactly the same as before.

@aarmey do we want the covariance signals (`covars`) to be normalized as well?

**Comment by aarmey** [comment]
You tell me—would it be better to have the scaling stored as separate weights, or in `covars`?

**Comment by andrewram4287** [comment]
> You tell me—would it be better to have the scaling stored as separate weights, or in `covars`?

I feel like it would be better to keep in separate weights, since we want to compare `covars` for different rankings, and it would easier to interpret with the weights stored separately

File: gmm/figures/figure6.py [hunk 7]
```diff
@@ -202,4 +162,109 @@ def make_blob_tensor(blob_DF):
...
+            ax[i + 1].tick_params(axis="y", rotation=0)
+            ax[i + 1].set_title("Mean Factors")
+
+    return
+
+
+def plotCovFactors(fac, facXA, DimCol, n_cluster, ax):
+    """Plots covarinace factors for dimension, time, and cluster"""
+    for i in range(fac.covars.shape[2]):
+        dff = pd.DataFrame(
```
Comment (aarmey) [inline gmm/figures/figure6.py:240]: Why doesn't this use `get_covariances_xarray()`?

**Review by aarmey — CHANGES_REQUESTED** [review]
Weights are now separate. One comment.

File: gmm/figures/figure6.py (hunk 7 above)
Comment (andrewram4287) [reply gmm/figures/figure6.py:240]: @aarmey `get_covariances_xarray()` rebuilds the tensor based on all the covariance factors. Instead we want to plot the covariance factors

Comment (aarmey) [reply gmm/figures/figure6.py:240]: Oh I see. Ok, we can keep this, but should probably make a `get_covariance_factors()` at some point.

If this builds we can merge?

Comment (andrewram4287) [reply gmm/figures/figure6.py:240]: @aarmey I will do that next. Yes looks like we can merge!
